compiler/computation_order/policy_gt.cc

```diff
@@ -0,0 +1,387 @@
+    }
+    return seq;
+}
+
+std::vector<Order> ComputeOrder(const Graph& graph, const SimpleGraph& sg, const std::vector<NodeSet>& seq) {
+    CHECK_GT(seq.size(), 0) << "seq should be non-empty";
+
+    std::map<Value*, size_t> block_index;
+    for (size_t i = 0; i + 1 < seq.size(); ++i) {
+        NodeSet set = SetMinus(seq[i + 1], seq[i]);
+        for (size_t k = 0; k < sg.n; ++k) {
+            if (!set[k]) continue;
+            block_index.insert({sg.value_list[k], i});
+        }
+    }
+
+    const std::vector<Node*> sorted = graph.GetTopologicallySortedNodes();
+    const size_t len = seq.size() - 1;
+    std::vector<std::vector<Value*>> blocks(len);
+    for (Node* node : sorted) {
+        for (Value* value : node->outputs()) {
+            const size_t idx = block_index[value];
+            blocks[idx].push_back(value);
+        }
+    }
+
+    std::vector<NodeSet> forget_sets(len);
+    for (size_t i = 0; i < len; ++i) {
+        if (i + 1 == len) {
+            // Do not forget any values in the last block
+            forget_sets[i] = NodeSet(sg.n);
+            continue;
+        }
+        const NodeSet boundary = Boundary(sg, seq[i + 1]);
+        const NodeSet nonboundary = SetMinus(seq[i + 1], boundary);
+        forget_sets[i] = SetMinus(nonboundary, seq[i]);
+    }
+
+    std::vector<Order> orders;
+    // Forward part
+    for (size_t i = 0; i < len; ++i) {
+        // Compute as usual
+        for (Value* value : blocks[i]) {
+            Node* producer = value->producer();
+            CHECK(producer);
+            orders.emplace_back(Order::kComputeForward, producer, nullptr);
+        }
+        // Forget non-boundary values
+        for (size_t k = 0; k < sg.n; ++k) {
+            if (!forget_sets[i][k]) continue;
+            Value* forget_value = sg.value_list[k];
+            orders.emplace_back(Order::kForgetForward, nullptr, forget_value);
+        }
+    }
+
+    // Backward part
+    for (size_t i = len; i-- > 0;) {
+        // Recompute forgotten values
+        for (Value* value : blocks[i]) {
+            const size_t k = sg.value_ids.find(value)->second;
+            if (!forget_sets[i][k]) continue;
+
+            Node* producer = value->producer();
+            CHECK(producer);
+            orders.emplace_back(Order::kComputeForward, producer, nullptr);
+        }
+
+        // Backward as usual
+        for (auto it = blocks[i].rbegin(); it != blocks[i].rend(); it++) {
+            Node* producer = (*it)->producer();
+            CHECK(producer);
+            orders.emplace_back(Order::kComputeBackward, producer, nullptr);
+        }
+    }
+
+    return orders;
+}
+
+int64_t AutomaticBudgetDetection() {
+#ifdef CHAINER_COMPILER_ENABLE_CUDA
+    size_t free, total;
+    if (cudaMemGetInfo(&free, &total) != cudaSuccess) {
+        return 100000000000LL;
+    }
+    const size_t used = total - free;
+    return total - 3 * used;
+#else
+    return nonstd::nullopt;
```

Comment:
```suggestion
    return absl::nullopt;
```
We are using absl like chainerx now.

Reply:
Thank you, fixed.